File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def solution(begin, target, words):
    if target not in set(words): return 0
    visited = set()
    queue = deque([(begin,0)])
    while queue:
        word, cnt = queue.popleft()
        if word== target:return cnt
        for w in words:
            dif = 0
            for idx, s in enumerate(word):
                if s != w[idx] : dif += 1
            if dif == 1 and w not in visited:
                queue.append((w,cnt+1))
                visited.add(w)

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
    public int solution(String begin, String target, String[] words) {
        if (!Arrays.asList(words).contains(target)) return 0;

        Queue<Word> queue = new ArrayDeque<>();
        Set<String> visited = new HashSet<>();

        queue.add(new Word(begin, 0));
        visited.add(begin);

        while (!queue.isEmpty()) {
            Word current = queue.poll();

            if (current.word.equals(target)) {
                return current.depth;
            }

            for (String next : words) {
                if (!visited.contains(next) && canTransform(current.word, next)) {
                    visited.add(next);
                    queue.add(new Word(next, current.depth + 1));
                }
            }
        }

        return 0;
    }

    static boolean canTransform(String a, String b) {
        int diff = 0;
        for (int i = 0; i < a.length(); i++) {
            if (a.charAt(i) != b.charAt(i)) diff++;
        }
        return diff == 1;
    }

    static class Word {
        String word;
        int depth;

        Word(String word, int depth) {
            this.word = word;
            this.depth = depth;
        }
    }
}